lightning-invoice/src/lib.rs

```diff
@@ -901,6 +906,79 @@ impl Deref for PositiveTimestamp {
 }
 
 impl Invoice {
+	/// Utility to construct an invoice.
+	pub fn from_channelmanager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>(
+		channelmanager: &ChannelManager<Signer, M, T, K, F, L>, amt_msat: Option<u64>, description: String,
+		network: Network,
+	) -> Result<Invoice, CreationError>
+	where
+		M::Target: chain::Watch<Signer>,
+		T::Target: BroadcasterInterface,
+		K::Target: KeysInterface<Signer = Signer>,
+		F::Target: FeeEstimator,
+		L::Target: Logger,
+	{
+		// Marshall route hints.
+		let our_channels = channelmanager.list_usable_channels();
+		let mut route_hints = vec![];
+		let mut min_final_cltv_expiry = 18;
+		for channel in our_channels {
+			let short_channel_id = match channel.short_channel_id {
+				Some(id) => id,
+				None => continue,
+			};
+			let forwarding_info = match channel.counterparty_forwarding_info {
+				Some(info) => info,
+				None => continue,
+			};
+			if forwarding_info.cltv_expiry_delta > min_final_cltv_expiry {
+				min_final_cltv_expiry = forwarding_info.cltv_expiry_delta;
+			}
+			route_hints.push(vec![RouteHintHop {
+				src_node_id: channel.remote_network_id,
+				short_channel_id,
+				fees: RoutingFees {
+					base_msat: forwarding_info.fee_base_msat,
+					proportional_millionths: forwarding_info.fee_proportional_millionths,
+				},
+				cltv_expiry_delta: forwarding_info.cltv_expiry_delta,
+				htlc_minimum_msat: None,
+				htlc_maximum_msat: None,
+			}]);
+		}
+
+		let (payment_hash, payment_secret) = channelmanager.create_inbound_payment(
+			amt_msat,
+			min_final_cltv_expiry as u32 * 10 * 60, // num blocks * 10 minutes per block
+			0,
+		);
+		let secp_ctx = Secp256k1::new();
+		let our_node_pubkey = channelmanager.get_our_node_id();
+		let mut invoice = InvoiceBuilder::new(match network {
+			Network::Bitcoin => Currency::Bitcoin,
+			Network::Testnet => Currency::BitcoinTestnet,
+			Network::Regtest => Currency::Regtest,
+			Network::Signet => panic!("Signet invoices not supported"),
+		})
+		.description(description)
+		.current_timestamp()
+		.payee_pub_key(our_node_pubkey)
+		.payment_hash(Hash::from_slice(&payment_hash.0).unwrap())
+		.payment_secret(payment_secret)
+		.features(InvoiceFeatures::known())
+		.min_final_cltv_expiry(min_final_cltv_expiry.into());
+		if let Some(amt) = amt_msat {
+			invoice = invoice.amount_pico_btc(amt * 10);
+		}
+		for hint in route_hints.drain(..) {
+			invoice = invoice.route(hint);
+		}
+
+		invoice.build_signed(|msg_hash| {
+			secp_ctx.sign_recoverable(msg_hash, &channelmanager.get_our_node_secret())
```

Comment:
Rather than exposing the secret, maybe `ChannelManager` could provide an interface for signing?

Reply:
If we sign anything blindly, it feels like basically the same thing - we'd have to actually have logic knowing what we're signing to avoid it, which would be great, except then we get into lightning depending on lightning-invoice and circular dependencies :/.

Reply:
Pretty agree with Jeff here, does isolating the utility in its own module resolve the circular dependency ?

Extracting the node secret key from the node will let any malicious entity signs a bunch of random invoices with known hash/preimages. If the victim node is a merchant it opens the way to further social attacks, like contacting the support to claim that a service/good is owed and has not been delivered ?

Reply:
> does isolating the utility in its own module resolve the circular dependency

No, the only solution we could do would be to pull at least `Features` (and probably a few other tiny things) into a new crate which both rust-lightning and lightning-invoice depend on. I'm not super excited about that, but its possible.

> Extracting the node secret key from the node will let any malicious entity signs a bunch of random invoices with known hash/preimages

Yea, we don't really have a concrete story for the node secret's security model right now. We do require it in ChannelManager, so we have to assume anyone with access to the `ChannelManager` can, in principle, do anything they want with the node secret (at least its on the same host/in the same memory space). Its unclear if there's a lot we can do about that, either, we need it to read onions. We could make reading onions a part of the keys interface, but that does seem like an API that is going to change a *ton*.

In general, I think we really need a spec update to let us sign invoices with a different key than onion decryption (etc). cc @devrandom.

Reply:
It would be good not to go further in the direction of passing around the private key.

I think we can have a function `sign_invoice` on `KeysInterface` with the following arguments:

- hrp
- data

and they could be either `Vec<u5>` or `Vec<u8>`.

If the developer actually wants to validate inside `KeysInterface`, they will have to re-parse the invoice.  I think that's fine, since you likely go across process boundaries when signing in the general case (hardware wallet / HSM / hardened signing server).

As to the node secret in other contexts - I think it boils down to signing things and computing shared secrets.  I think both of these can be behind `KeysInterface` (or a smaller trait that can be gotten from it).  I'm happy to try implementing there's doubt.